File: api/backend/pkg/api/invite.go
package api

import (
	"database/sql"
	"errors"
	"github.com/dokedu/dokedu-api-beta/services/pkg/api/status"
	"github.com/dokedu/dokedu-api-beta/services/pkg/helper"
	"github.com/dokedu/dokedu-api-beta/services/pkg/models"
	"github.com/dokedu/dokedu-api-beta/services/pkg/modules"
	"github.com/labstack/echo/v4"
	"net/http"
)

type InviteUserRequest struct {
	Email          string `validate:"required,email" json:"email"`
	FirstName      string `validate:"required" json:"first_name"`
	LastName       string `validate:"required" json:"last_name"`
	OrganisationID string `validate:"required" json:"organisation_id"`
	Role           string `validate:"required,oneof=admin teacher teacher_guest" json:"role"`
}

func InviteUser(mods *modules.Modules) func(c echo.Context) error {
	return func(c echo.Context) error {
		ctx := c.(helper.ApiContext)

		var req InviteUserRequest
		if err := ctx.Bind(&req); err != nil {
			return echo.NewHTTPError(http.StatusBadRequest, status.ErrBadRequest)
		}

		// validate request
		if err := ctx.Validate(req); err != nil {
			return echo.NewHTTPError(http.StatusBadRequest, status.ErrBadRequest)
		}

		// check if user is admin or owner in the organisation
		cnt, err := ctx.Tx.NewSelect().
			TableExpr("accounts a").
			Join("inner join identities i on i.id = a.identity_id").
			Where("i.user_id = ?", ctx.Claims.Sub).
			Where("a.organisation_id = ?", req.OrganisationID).
			Where("a.role = 'admin' or role = 'owner'").
			Where("a.deleted_at IS NULL").
			Where("i.deleted_at IS NULL").
			Count(ctx.Request().Context())

		if err != nil {
			return err
		}

		if cnt != 1 {
			return echo.NewHTTPError(http.StatusForbidden, status.ErrForbidden)
		}

		// check if user already exists
		var user models.User
		err = ctx.Tx.NewSelect().
			TableExpr("\"auth\".users").
			Column("id", "email").
			Where("email = ?", req.Email).
			Scan(ctx.Request().Context(), &user)

		if errors.Is(err, sql.ErrNoRows) {
			user, err = mods.Supabase.InviteUserByEmail(req.Email)
			if err != nil {
				return err
			}
		} else if err != nil {
			return err
		}

		// check if the user has an identity already
		var identity models.Identity
		err = ctx.Tx.NewSelect().
			Model(&identity).
			Where("user_id = ?", user.ID).
			Where("deleted_at IS NULL").
			Scan(ctx.Request().Context())
		if err != nil {
			if !errors.Is(err, sql.ErrNoRows) {
				return err
			}

			// create identity
			identity.UserID = user.ID
			identity.GlobalRole = "default"

			_, err = ctx.Tx.NewInsert().
				Model(&identity).
				Exec(ctx.Request().Context())
			if err != nil {
				return err
			}
		}

		// check if account already exists
		cnt, err = ctx.Tx.NewSelect().
			Table("accounts").
			Where("identity_id = ?", identity).
			Where("organisation_id = ?", req.OrganisationID).
			Where("deleted_at IS NULL").
			Count(ctx.Request().Context())
		if err != nil {
			return err
		}

		if cnt != 0 {
			return echo.NewHTTPError(http.StatusConflict, "user already part of organisation")
		}

		// create account
		account := models.Account{
			IdentityID: sql.NullString{
				String: identity.ID,
				Valid:  true,
			},
			OrganisationID: req.OrganisationID,
			Role:           req.Role,
			FirstName:      req.FirstName,
			LastName:       req.LastName,
		}
		_, err = ctx.Tx.NewInsert().
			Model(&account).
			Exec(ctx.Request().Context())

		if err != nil {
			return err
		}

		return ctx.JSON(http.StatusOK, status.ResOK)
	}
}

File: api/functions/users/invite.ts
export function InviteUser(mods: modules.Modules) {
    return async (c: echo.Context) => {
        const ctx = c as helper.ApiContext;

        const req = await ctx.bind<InviteUserRequest>();

        // validate request
        if (!ctx.validate(req)) {
            return ctx.badRequest();
        }

        // check if user is admin or owner in the organisation
        const cnt = await ctx.tx.newSelect()
            .from("accounts as a")
            .innerJoin("identities as i").on("i.id = a.identity_id")
            .where("i.user_id = ?", ctx.claims.sub)
            .where("a.organisation_id = ?", req.organisationID)
            .where("a.role = 'admin' or role = 'owner'")
            .where("a.deleted_at IS NULL")
            .where("i.deleted_at IS NULL")
            .count(ctx.request().context);

        if (cnt !== 1) {
            return ctx.forbidden();
        }

        // check if user already exists
        let user = await ctx.tx.newSelect()
            .from("auth.users")
            .column("id", "email")
            .where("email = ?", req.email)
            .fetchOne<models.User>(ctx.request().context);

        if (user === undefined) {
            user = await mods.supabase.inviteUserByEmail(req.email);
        }

        // check if the user has an identity already
        let identity = await ctx.tx.newSelect()
            .from("identities")
            .where("user_id = ?", user.id)
            .where("deleted_at IS NULL")
            .fetchOne<models.Identity>(ctx.request().context);
        if (identity === undefined) {
            // create identity
            identity = {
                userID: user.id,
                globalRole: "default",
            };

            await ctx.tx.newInsert()
                .into("identities")
                .values(identity)
                .execute(ctx.request().context);
        }

        // check if account already exists
        const cnt = await ctx.tx.newSelect()
            .from("accounts")
            .where("identity_id = ?", identity.id)
            .where("organisation_id = ?", req.organisationID)
            .where("deleted_at IS NULL")
            .count(ctx.request().context);

        if (cnt !== 0) {
            return ctx.conflict("user already part of organisation");
        }

        // create account
        const account = {
            identityID: {
                string: identity.id,
                valid: true,
            },
            organisationID: req.organisationID,
            role: req.role,
            firstName: req.firstName,
            lastName: req.lastName,
        };

        await ctx.tx.newInsert()
            .into("accounts")
            .values(account)
            .execute(ctx.request().context);

        return ctx.ok();
    };
}